### PR TITLE
ENH: plus attrs now are adjusted to be in modulo of step

### PR DESCRIPTION
**Pull request by KatherineCaley**
## Summary by Sourcery

Enhance sequence slicing by adjusting plus attributes to be in modulo of step, ensuring consistent handling of sequence views. Add tests to verify correct behavior when stop values do not align with step.

Enhancements:
- Adjust sequence slicing to ensure plus attributes are in modulo of step, improving consistency in sequence handling.

Tests:
- Add tests for sequence view attributes when stop value does not align with step, ensuring correct plus_start and plus_stop values.

### CODE REVIEW

File: src/cogent3/core/new_sequence.py [hunk 1]
```diff
@@ -2113,9 +2113,17 @@ def offset(self, value: int):
     def plus_start(self) -> int:
```
[the remark this answers is not in the document]
Comment (KatherineCaley) [reply src/cogent3/core/new_sequence.py:2113]: These will not produce the correct values in many cases

File: src/cogent3/core/new_alignment.py [hunk 2]
```diff
@@ -390,21 +384,46 @@ def get_seq_length(self, seqid: str) -> int:
...
+        seqid: str,
+        start: OptInt = None,
+        stop: OptInt = None,
+        step: OptInt = None,
     ) -> numpy.ndarray:
-        return self._data[seqid][start:stop]
+        start = start or 0
+        stop = stop if stop is not None else self.get_seq_length(seqid)
+        step = step or 1
+
```
Comment (GavinHuttley) [inline src/cogent3/core/new_alignment.py:397]: add assertion to check all components are >= 0. Same for the alignment variant

File: src/cogent3/core/new_sequence.py [hunk 3]
```diff
@@ -2113,9 +2113,17 @@ def offset(self, value: int):
...
+            # to positive indices, we add the length of the parent sequence to
+            # the stop. Note that when abs(self.step) > 1, we instead calculate
+            # the "true stop", i.e., the index that immediately follows (in the
+            # direction of the step) the last selected index in the slice. We then
+            # add the abs(self.step) to account for the fact that the stop uses
+            # exclusive indexing, and the start is inclusive.
+
             assert self.stop < 0, "expected stop on reverse strand SeqView < 0"
-            start = self.stop + self.parent_len + 1
+            stop = self.stop if self.step == -1 else self.start + len(self) * self.step
```
Comment (GavinHuttley) [inline src/cogent3/core/new_sequence.py:2125]: use `self.is_reversed`

File: tests/test_core/test_new_alignment.py [hunk 4]
```diff
@@ -623,6 +623,30 @@ def test_make_seqs_from_sequences(mk_cls):
...
+    )
+    seq_view_2 = new_sequence.SeqView(
+        parent="TTTT", parent_len=4, seqid="parent_2", alphabet=dna_alphabet
+    )
+
+    seq_1 = new_moltype.DNA.make_seq(seq=seq_view_1, name="seq_1")
+    seq_2 = new_moltype.DNA.make_seq(seq=seq_view_2, name="seq_2")
+
+    seqs = mk_cls([seq_1, seq_2], moltype="dna")
+    assert seqs.names == ["seq_1", "seq_2"]
```
Comment (GavinHuttley) [inline tests/test_core/test_new_alignment.py:646]: make these sets so the test is not sensitive to name order

File: tests/test_core/test_new_alignment.py [hunk 5]
```diff
@@ -1047,7 +1089,21 @@ def test_sequence_collection_degap(mk_cls, gap_ambig_seqs, rced):
...
     # Test empty sequences case
     empty_seqs = mk_cls({"empty1": "", "empty2": ""}, moltype="dna")
-    got_empty = empty_seqs.degap().to_dict()
+    empty_seqs = empty_seqs.degap()
+    got_empty = empty_seqs.to_dict()
+    expect_empty = {"empty1": "", "empty2": ""}
+    assert got_empty == expect_empty
+
+
+def test_sequence_collection_degap_one_off():
```
Comment (GavinHuttley) [inline tests/test_core/test_new_alignment.py:1098]: possibly don't need this test now?

**Review by GavinHuttley — APPROVED** [review]
💪 couple of small tweaks. Merge when you address those.
new_alignment line 3733, use `slice_record.is_reversed`

File: tests/test_core/test_new_alignment.py (hunk 5 above)
Comment (KatherineCaley) [reply tests/test_core/test_new_alignment.py:1098]: good catch!

File: src/cogent3/core/new_sequence.py (hunk 3 above)
Comment (KatherineCaley) [reply src/cogent3/core/new_sequence.py:2125]: @GavinHuttley  I'm actually not trying to check if the slice record is reversed, we are already within a conditional where we know we are reversed. I am checking to see if the step is -1, which means we can avoid calculating the modulo-adjusted stop

**Pull request merged**